File: legacy/python/src/templates.py
"""Template registry: the 10 supported bot types + request matching."""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from . import db

# (fa-normalized keywords, weight) — first match wins per template, then max score.
# fa-normalized: letters lowercased, ي->ی, ك->ک, zero-width/arabic diacritics stripped.


def _norm(text: str) -> str:
    t = text.lower()
    t = t.replace("ي", "ی").replace("ك", "ک").replace("ۀ", "ه").replace("ة", "ه")
    t = re.sub(r"[\u200c\u200b\u200d\u200e\u200f\u064b-\u0652]", "", t)
    return t


@dataclass
class Template:
    id: str
    name: str
    desc: str
    keywords: list[str]           # fa-normalized
    features: dict                 # feature -> bool
    base_config: dict = field(default_factory=dict)
    needs_owner_id: bool = True   # most bots need an admin id
    setup_hint: str = ""


TEMPLATES: list[Template] = [
    Template(
        id="autoreply",
        name="ربات پاسخگوی خودکار",
        desc="به پیام‌های کاربران با کلمات کلیدی و متن‌های از پیش تعیین‌شده جواب می‌ده.",
        keywords=["پاسخ", "جواب", "اتورپلی", "اتو ریپلای", "اتو", "چت", "گفتگو", "مشاوره", "reply", "answer", "auto"],
        features={"autoreply": True, "shop": False, "joiner": False, "groupadmin": False,
                  "broadcast": False, "welcome": False, "poll": False, "antispam": False, "forward": False},
        base_config={"default_reply": "سلام! چطور می‌تونم کمکت کنم؟"},
        needs_owner_id=False,
        setup_hint="بعد از ساخت، /panel را بزن و از «پاسخ خودکار» کلمات کلیدی را اضافه کن.",
    ),
    Template(
        id="shop",
        name="ربات فروشگاه / درگاه",
        desc="فروشگاه با محصولات، دکمه خرید و پرداخت با ستاره‌های تلگرام (Stars).",
        keywords=["فروشگاه", "فروش", "خرید", "درگاه", "محصول", "پرداخت", "شارژ", "shop", "store", "buy", "payment", "price"],
        features={"autoreply": False, "shop": True, "joiner": False, "groupadmin": False,
                  "broadcast": True, "welcome": False, "poll": False, "antispam": False, "forward": False},
        needs_owner_id=True,
        setup_hint="بعد از ساخت، /panel → فروشگاه: محصولات را اضافه کن. قیمت‌ها با ستاره‌ی تلگرام پرداخت می‌شوند.",
    ),
    Template(
        id="joiner",
        name="ربات جوینر / ممبرگیر",
        desc="کاربران را مجبور می‌کند اول عضو کانال/گروه تو شوند تا از ربات استفاده کنند.",
        keywords=["جوین", "ممبر", "عضو", "عضویت", "کانال", "join", "member", "subscribe"],
        features={"autoreply": False, "shop": False, "joiner": True, "groupadmin": False,
                  "broadcast": False, "welcome": False, "poll": False, "antispam": False, "forward": False},
        needs_owner_id=True,
        setup_hint="بعد از ساخت، /panel → گیت عضویت: آیدی کانال را بده و ربات را ادمین آن کانال کن.",
    ),
    Template(
        id="groupadmin",
        name="ربات مدیریت گروه",
        desc="کیک، بن، میوت، آنتی‌لینک و ضداسپم برای گروه خودت.",
        keywords=["ادمین", "گروه", "کیک", "میت", "مدیریت", "mod", "admin", "kick", "ban"],
        features={"autoreply": False, "shop": False, "joiner": False, "groupadmin": True,
                  "broadcast": False, "welcome": True, "poll": False, "antispam": True, "forward": False},
        needs_owner_id=True,
        setup_hint="ربات را ادمین گروهت کن؛ دستورات /kick /ban /mute /antilink را در گروه اجرا کن.",
    ),
    Template(
        id="broadcast",
        name="ربات اطلاع‌رسانی / برادکست",
        desc="ارسال پیام همگانی به همه‌ی کسانی که ربات را استارت کرده‌اند.",
        keywords=["اطلاع", "خبر", "برادکست", "همگانی", "اعلام", "کانال", "broadcast", "announce", "news"],
        features={"autoreply": True, "shop": False, "joiner": False, "groupadmin": False,
                  "broadcast": True, "welcome": False, "poll": False, "antispam": False, "forward": False},
        base_config={"default_reply": "برای اطلاع‌رسانی با ادمین در ارتباط باشید."},
        needs_owner_id=True,
        setup_hint="بعد از ساخت، /broadcast <متن> را بفرست تا به همه‌ی کاربران برسد.",
    ),
    Template(
        id="welcome",
        name="ربات خوش‌آمدگویی",
        desc="به اعضای جدید گروه پیام خوش‌آمد بده و گروه را حرفه‌ای نشان بده.",
        keywords=["خوش امد", "خوشامد", "ویلکام", "welcome", "ورود"],
        features={"autoreply": False, "shop": False, "joiner": False, "groupadmin": False,
                  "broadcast": False, "welcome": True, "poll": False, "antispam": False, "forward": False},
        needs_owner_id=True,
        setup_hint="بعد از ساخت، /setwelcome <متن> را بزن و ربات را ادمین گروهت کن.",
    ),
    Template(
        id="poll",
        name="ربات نظرسنجی",
        desc="ساخت نظرسنجی با دستور ساده و مشاهده نتایج.",
        keywords=["نظرسنجی", "رای", "نظر", "poll", "vote", "survey"],
        features={"autoreply": False, "shop": False, "joiner": False, "groupadmin": False,
                  "broadcast": False, "welcome": False, "poll": True, "antispam": False, "forward": False},
        needs_owner_id=True,
        setup_hint="بعد از ساخت: /poll سوال|گزینه۱|گزینه۲|...",
    ),
    Template(
        id="antispam",
        name="ربات ضداسپم",
        desc="حذف خودکار پیام‌های حاوی لینک و کلمات اسپم در گروه.",
        keywords=["اسپم", "ضد اسپم", "لینک", "هرز", "spam", "antispam"],
        features={"autoreply": False, "shop": False, "joiner": False, "groupadmin": False,
                  "broadcast": False, "welcome": False, "poll": False, "antispam": True, "forward": False},
        needs_owner_id=True,
        setup_hint="بعد از ساخت ربات را ادمین گروه کن؛ /antilink on را بزن.",
    ),
    Template(
        id="card",
        name="ربات کارت ویزیت / معرفی",
        desc="به هر کسی که استارت کند، معرفی‌نامه‌ی تو را نشان می‌دهد.",
        keywords=["کارت", "ویزیت", "معرفی", "رزومه", "بیو", "card", "bio", "introduce"],
        features={"autoreply": True, "shop": False, "joiner": False, "groupadmin": False,
                  "broadcast": False, "welcome": False, "poll": False, "antispam": False, "forward": False},
        base_config={
            "default_reply": "سلام! 👋\nمن ربات معرفی هستم.\nبرای مشاهده‌ی اطلاعات، /info را بزن.",
        },
        needs_owner_id=False,
        setup_hint="بعد از ساخت /setinfo <متن معرفی> را بزن.",
    ),
    Template(
        id="forward",
        name="ربات فوروارد خودکار",
        desc="پیام‌های کانال/گروه مبدأ را خودکار به مقصد فوروارد می‌کند.",
        keywords=["فوروارد", "انتقال", "کپی", "forward", "copy", "relay"],
        features={"autoreply": False, "shop": False, "joiner": False, "groupadmin": False,
                  "broadcast": False, "welcome": False, "poll": False, "antispam": False, "forward": True},
        needs_owner_id=True,
        setup_hint="بعد از ساخت /setforward @مبدا|@مقصد را بزن (ربات در هر دو ادمین باشد).",
    ),
]


def match_request(text: str) -> tuple[Template | None, int]:
    """Return (best_template, score). Score 0 = no match."""
    norm = _norm(text)
    best: Template | None = None
    best_score = 0
    for t in TEMPLATES:
        score = 0
        for kw in t.keywords:
            nk = _norm(kw)
            if nk in norm:
                score += len(nk)  # longer keyword = stronger signal
        if score > best_score:
            best = t
            best_score = score
    return best, best_score


def template_features(t: Template) -> dict:
    f = {k: False for k in db.default_config()["features"]}
    f.update(t.features)
    return f


def base_config_for(t: Template) -> dict:
    cfg = db.default_config()
    cfg["features"] = template_features(t)
    for k, v in t.base_config.items():
        cfg[k] = v
    return cfg

File: src/templates.ts
/** Template registry: the 10 supported bot types + Persian/English request matching.
 *  Ported from the Python prototype (legacy/python/src/templates.py). */

export type FeatureId =
  | "autoreply"
  | "shop"
  | "joiner"
  | "groupadmin"
  | "broadcast"
  | "welcome"
  | "poll"
  | "antispam"
  | "card"
  | "forward"
  | "content";

export const FEATURE_IDS: FeatureId[] = [
  "autoreply",
  "shop",
  "joiner",
  "groupadmin",
  "broadcast",
  "welcome",
  "poll",
  "antispam",
  "card",
  "forward",
  "content",
];

export interface Template {
  id: string;
  name: string;
  desc: string;
  keywords: string[]; // fa-normalized
  features: Record<FeatureId, boolean>;
  baseConfig: Record<string, unknown>;
  needsOwnerId: boolean;
  setupHint: string;
}

/** fa-normalization: lowercase, ي->ی, ك->ک, آ->ا, strip zero-width chars + Arabic diacritics. */
export function norm(text: string): string {
  let t = text.toLowerCase();
  t = t.replace(/ي/g, "ی").replace(/ك/g, "ک").replace(/ۀ/g, "ه").replace(/ة/g, "ه").replace(/آ/g, "ا");
  t = t.replace(/[\u200c\u200b\u200d\u200e\u200f\u064b-\u0652]/g, "");
  return t;
}

export const TEMPLATES: Template[] = [
  {
    id: "autoreply",
    name: "ربات پاسخگوی خودکار",
    desc: "به پیام‌های کاربران با کلمات کلیدی و متن‌های از پیش تعیین‌شده جواب می‌دهد.",
    keywords: ["پاسخ", "جواب", "اتورپلی", "اتو ریپلای", "اتو", "چت", "گفتگو", "مشاوره", "reply", "answer", "auto"],
    features: { autoreply: true, shop: false, joiner: false, groupadmin: false, broadcast: false, welcome: false, poll: false, antispam: false, card: false, forward: false, content: false },
    baseConfig: { defaultReply: "سلام! چطور می‌تونم کمکت کنم؟" },
    needsOwnerId: false,
    setupHint: "بعد از ساخت، /panel را بزن و از «پاسخ خودکار» کلمات کلیدی را اضافه کن.",
  },
  {
    id: "shop",
    name: "ربات فروشگاه / درگاه",
    desc: "فروشگاه با محصولات، دکمه خرید و پرداخت با ستاره‌های تلگرام (Stars).",
    keywords: ["فروشگاه", "فروش", "خرید", "درگاه", "محصول", "پرداخت", "شارژ", "shop", "store", "buy", "payment", "price"],
    features: { autoreply: false, shop: true, joiner: false, groupadmin: false, broadcast: true, welcome: false, poll: false, antispam: false, card: false, forward: false, content: false },
    baseConfig: {},
    needsOwnerId: true,
    setupHint: "بعد از ساخت، /panel → فروشگاه: محصولات را اضافه کن. قیمت‌ها با ستاره‌ی تلگرام پرداخت می‌شوند.",
  },
  {
    id: "joiner",
    name: "ربات جوینر / ممبرگیر",
    desc: "کاربران را مجبور می‌کند اول عضو کانال/گروه تو شوند تا از ربات استفاده کنند.",
    keywords: ["جوین", "ممبر", "عضو", "عضویت", "کانال", "join", "member", "subscribe"],
    features: { autoreply: false, shop: false, joiner: true, groupadmin: false, broadcast: false, welcome: false, poll: false, antispam: false, card: false, forward: false, content: false },
    baseConfig: {},
    needsOwnerId: true,
    setupHint: "بعد از ساخت، /panel → گیت عضویت: آیدی کانال را بده و ربات را ادمین آن کانال کن.",
  },
  {
    id: "groupadmin",
    name: "ربات مدیریت گروه",
    desc: "کیک، بن، میوت، آنتی‌لینک و ضداسپم برای گروه خودت.",
    keywords: ["ادمین", "گروه", "کیک", "میت", "مدیریت", "mod", "admin", "kick", "ban"],
    features: { autoreply: false, shop: false, joiner: false, groupadmin: true, broadcast: false, welcome: true, poll: false, antispam: true, card: false, forward: false, content: false },
    baseConfig: {},
    needsOwnerId: true,
    setupHint: "ربات را ادمین گروهت کن؛ دستورات /kick /ban /mute /antilink را در گروه اجرا کن.",
  },
  {
    id: "broadcast",
    name: "ربات اطلاع‌رسانی / برادکست",
    desc: "ارسال پیام همگانی به همه‌ی کسانی که ربات را استارت کرده‌اند.",
    keywords: ["اطلاع", "خبر", "برادکست", "همگانی", "اعلام", "کانال", "broadcast", "announce", "news"],
    features: { autoreply: true, shop: false, joiner: false, groupadmin: false, broadcast: true, welcome: false, poll: false, antispam: false, card: false, forward: false, content: false },
    baseConfig: { defaultReply: "برای اطلاع‌رسانی با ادمین در ارتباط باشید." },
    needsOwnerId: true,
    setupHint: "بعد از ساخت، /broadcast <متن> را بفرست تا به همه‌ی کاربران برسد.",
  },
  {
    id: "welcome",
    name: "ربات خوش‌آمدگویی",
    desc: "به اعضای جدید گروه پیام خوش‌آمد بده و گروه را حرفه‌ای نشان بده.",
    keywords: ["خوش امد", "خوشامد", "ویلکام", "welcome", "ورود"],
    features: { autoreply: false, shop: false, joiner: false, groupadmin: false, broadcast: false, welcome: true, poll: false, antispam: false, card: false, forward: false, content: false },
    baseConfig: {},
    needsOwnerId: true,
    setupHint: "بعد از ساخت، /setwelcome <متن> را بزن و ربات را ادمین گروهت کن.",
  },
  {
    id: "poll",
    name: "ربات نظرسنجی",
    desc: "ساخت نظرسنجی با دستور ساده و مشاهده نتایج.",
    keywords: ["نظرسنجی", "رای", "نظر", "poll", "vote", "survey"],
    features: { autoreply: false, shop: false, joiner: false, groupadmin: false, broadcast: false, welcome: false, poll: true, antispam: false, card: false, forward: false, content: false },
    baseConfig: {},
    needsOwnerId: true,
    setupHint: "بعد از ساخت: /poll سوال|گزینه۱|گزینه۲|...",
  },
  {
    id: "antispam",
    name: "ربات ضداسپم",
    desc: "حذف خودکار پیام‌های حاوی لینک و کلمات اسپم در گروه.",
    keywords: ["اسپم", "ضد اسپم", "لینک", "هرز", "spam", "antispam"],
    features: { autoreply: false, shop: false, joiner: false, groupadmin: false, broadcast: false, welcome: false, poll: false, antispam: true, card: false, forward: false, content: false },
    baseConfig: {},
    needsOwnerId: true,
    setupHint: "بعد از ساخت ربات را ادمین گروه کن؛ /antilink on را بزن.",
  },
  {
    id: "card",
    name: "ربات کارت ویزیت / معرفی",
    desc: "به هر کسی که استارت کند، معرفی‌نامه‌ی تو را نشان می‌دهد.",
    keywords: ["کارت", "ویزیت", "معرفی", "رزومه", "بیو", "card", "bio", "introduce"],
    features: { autoreply: true, shop: false, joiner: false, groupadmin: false, broadcast: false, welcome: false, poll: false, antispam: false, card: true, forward: false, content: false },
    baseConfig: { defaultReply: "سلام! 👋\nمن ربات معرفی هستم.\nبرای مشاهده‌ی اطلاعات، /info را بزن." },
    needsOwnerId: false,
    setupHint: "بعد از ساخت /setinfo <متن معرفی> را بزن.",
  },
  {
    id: "forward",
    name: "ربات فوروارد خودکار",
    desc: "پیام‌های کانال/گروه مبدأ را خودکار به مقصد فوروارد می‌کند.",
    keywords: ["فوروارد", "انتقال", "کپی", "forward", "copy", "relay"],
    features: { autoreply: false, shop: false, joiner: false, groupadmin: false, broadcast: false, welcome: false, poll: false, antispam: false, card: false, forward: true, content: false },
    baseConfig: {},
    needsOwnerId: true,
    setupHint: "بعد از ساخت /setforward @مبدا|@مقصد را بزن (ربات در هر دو ادمین باشد).",
  },
  {
    id: "content",
    name: "ربات جمع‌آوری محتوا",
    desc: "اخبار AI و لیست APIهای رایگان رو جمع می‌کند (RSS + publicapis.org — کاملاً رایگان).",
    keywords: ["جمع", "محتوا", "خبر", "اخبار", "rss", "feed", "content", "collector", "news"],
    features: { autoreply: false, shop: false, joiner: false, groupadmin: false, broadcast: false, welcome: false, poll: false, antispam: false, card: false, forward: false, content: true },
    baseConfig: {},
    needsOwnerId: true,
    setupHint: "دستورات /news (اخبار AI) و /apis (APIهای رایگان) را بزن.",
  },
];

export function defaultFeatures(): Record<FeatureId, boolean> {
  const f = {} as Record<FeatureId, boolean>;
  for (const id of FEATURE_IDS) f[id] = false;
  return f;
}

export function templateById(id: string): Template | undefined {
  return TEMPLATES.find((t) => t.id === id);
}

export function baseConfigFor(t: Template): Record<string, unknown> {
  return { features: { ...defaultFeatures(), ...t.features }, ...t.baseConfig };
}

export interface MatchResult {
  template: Template | null;
  score: number;
}

/** Longer keyword = stronger signal. First match wins per template, then max total score. */
export function matchRequest(text: string): MatchResult {
  const n = norm(text);
  let best: Template | null = null;
  let bestScore = 0;
  for (const t of TEMPLATES) {
    let score = 0;
    for (const kw of t.keywords) {
      const nk = norm(kw);
      if (n.includes(nk)) score += nk.length;
    }
    if (score > bestScore) {
      best = t;
      bestScore = score;
    }
  }
  return { template: best, score: bestScore };
}

/** Simple heuristic: does this text look like a URL (for antilink)? */
export function hasLink(text: string): boolean {
  return /(https?:\/\/|www\.|t\.me\/|telegram\.me\/)[^\s]+/i.test(text);
}
